src/rules/unnecessaryElseRule.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * @license
+ * Copyright 2019 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+import { codeExamples } from "./code-examples/unnecessaryElse.examples";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        description: Lint.Utils.dedent`
+        Disallows \`else\` blocks following \`if\` blocks ending with a \`break\`, \`continue\`, \`return\`, or \`throw\` statement.`,
+        descriptionDetails: "",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+        When an \`if\` block is guaranteed to exit control flow when entered,
+        it is unnecessary to add an \`else\` statement.
+        The contents that would be in the \`else\` block can be placed after the end of the \`if\` block.`,
+        ruleName: "unnecessary-else",
+        type: "functionality",
+        typescriptOnly: false,
+        codeExamples,
+    };
+    /* tslint:disable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string): string {
+        return `The preceding \`if\` block ends with a \`${name}\` statement. This \`else\` block is unnecessary.`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+function walk(ctx: Lint.WalkContext<void>): void {
+    ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (utils.isIfStatement(node)) {
+            let jumpStatement;
+            if (
+                node.thenStatement.kind !== ts.SyntaxKind.Block &&
+                !utils.isSameLine(ctx.sourceFile, node.thenStatement.pos, node.thenStatement.end)
```

Comment:
I'm a little confused, is it actually necessary to check `utils.isSameLine`? `node` and its `.thenStatement` should behave the same regardless of whitespace position, right?

Reply:
I have verified this multiple times,
if() return;
else return;

behaves very differently from 

if()
	return;
else
	return;

which is why I had to check `utils.isSameLine`. Could you kindly suggest a way to optimize this ? @JoshuaKGoldberg

Reply:
Interesting! In what ways are you finding them to be different?

With this test, I'm getting the same result reported for both:

```ts
function testSameLine(condition) {
    if (condition) return 'a';
    else return 'b';
}

function testNextLine(condition) {
    if (condition)
    return 'a';
    else
    return 'b';
}

for (const condition of [true, false]) {
    console.log(condition);
    console.log("\tSame line:", testSameLine(condition));
    console.log("\tNext line:", testNextLine(condition));
    console.log("");
}
```

```
true
    Same line: a
    Same line: a

false
    Same line: b
    Same line: b
```

Reply:
If I remove any of the `utils.isSameLine` checks I get the below error and the rule doesnt work
![logs](https://user-images.githubusercontent.com/22490998/53304449-50e85300-389b-11e9-9b8c-834727767a51.png)

Reply:
Referring to the `TypeError; cannot read property '0' of undefined`?

It looks like the problem is from the `jumpStatement = getJumpStatement(getLastStatement(node.thenStatement as ts.Block));`. The rule can't assume `node.thenStatement as ts.Block` is valid because `node.thenStatement` might be some other kind of node, such as a `ReturnStatement` in the sample code in this thread.

You should be able to simplify the cases within the `if (utils.isIfStatement` block into just the following:

* `tsutils.isBlock(node.thenStatement)`
* Not ☝

...rather than the three cases you have now.

Reply:
Done https://github.com/palantir/tslint/pull/4502/commits/f36d4201e2fa8ec3aec60a022db2bf51435aceaf

Reply:
Done https://github.com/palantir/tslint/pull/4502/commits/4c29b4309e698117979d5bd9a8e62b9fbc337b1f